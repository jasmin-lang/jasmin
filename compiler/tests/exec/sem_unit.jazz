fn vpsllv_4u32(reg u128 x y) -> reg u128 {
  reg u128 z;
  z = #VPSLLV_4u32(x, y);
  return z;
}

fn vpsllv_8u32(reg u256 x y) -> reg u256 {
  reg u256 z;
  z = #VPSLLV_8u32(x, y);
  return z;
}

fn vpsllv_4u64(reg u256 x y) -> reg u256 {
  reg u256 z;
  z = #VPSLLV_4u64(x, y);
  return z;
}

fn shrd_16(reg u16 x y, reg u8 c) -> reg u16 {
  ?{}, x = #SHRD_16(x, y, c);
  return x;
}
