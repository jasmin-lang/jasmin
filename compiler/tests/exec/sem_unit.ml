open Jasmin
open Execlib

let w_of_z sz z = Values.Vword (sz, Conv.word_of_z sz z)
let w_of_string sz n = w_of_z sz (Z.of_string n)
let w8 i = w_of_string Wsize.U8 i
let w16 i = w_of_string Wsize.U16 i
let w128 i = w_of_string Wsize.U128 i
let w256 i = w_of_string Wsize.U256 i

let () =
  let prog = load_file "sem_unit.jazz" in
  exec prog [] "vpsllv_4u32" @@ List.map w128 [ "1"; "1000000" ];
  exec prog [] "vpsllv_4u32"
    [
      w128 "0x5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a5a";
      w128 "0x00000004000000030000000200000001";
    ];
  exec prog [] "vpsllv_8u32" @@ List.map w256 [ "1"; "1000000" ];
  exec prog [] "vpsllv_4u64"
    [
      w256 "0x00ffff001248a00000ffff001248a00000ffff001248a00000ffff001248a000";
      w256 "0x0000000000000002000000000000000400000000000000080000000000000010";
    ];
  exec prog [] "shrd_16" [ w16 "2"; w16 "9"; w8 "16" ];
  exec prog [] "shrd_16" [ w16 "2"; w16 "9"; w8 "17" ];
  ()
