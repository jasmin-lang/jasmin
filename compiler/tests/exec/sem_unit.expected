vpsllv_4u32(0X1; 0XF4240) = 0X0
vpsllv_4u32(0X5A5A5A5A5A5A5A5A5A5A5A5A5A5A5A5A; 0X4000000030000000200000001) = 0XA5A5A5A0D2D2D2D069696968B4B4B4B4
vpsllv_8u32(0X1; 0XF4240) = 0X0
vpsllv_4u64(0XFFFF001248A00000FFFF001248A00000FFFF001248A00000FFFF001248A000; 0X2000000000000000400000000000000080000000000000010) = 0X3FFFC00492280000FFFF001248A0000FFFF001248A00000FF001248A0000000
shrd_16(0X2; 0X9; 0X10) = 0X9
shrd_16(0X2; 0X9; 0X11) failed with arithmetic error
